### ISSUE TITLE
Upgrading FlashAttention to V3

### ISSUE THREAD

**Issue by drisspg** [pytorch/pytorch#148891]
# Summary

We are currently building and utilizing FlashAttention2 for torch.nn.functional.scaled_dot_product_attention
 
Up until recently the files we build and our integration was very manual. We recently changed this and made FA a third_party/submodule: https://github.com/pytorch/pytorch/pull/146372

This makes it easier to pull in new files (including those for FAv3) however due to the fact that third_party extensions do not have a mechanism to be re-integrated into ATen the build system + flash_api is still manual. 

### Plan
At a very high level we have a few options. I will for the sake of argument though not include the runtime dependency option.   So for know lets assume we need to build and ship the kernels in libtorchcuda.so

1. Replace entirely FAv2 w/ FAv3:
This up until recently seemed like a non ideal option since we would lose FA support for A100 + machines. This has changed in: https://github.com/Dao-AILab/flash-attention/commit/7bc3f031a40ffc7b198930b69cf21a4588b4d2f9 and therefor this seems like a much more viable option, and least binary size impactful. I think the main difference is that FAv3 doesn't support Dropout. TBD if this a large enough blocker.

2. Add FAv2 along w/ FAv2
This would require adding another backend to SDPA for FAv3. This would naively have a large impact to binary size, however we could choose to only build these kernels on H100 machines. 

I am personally in favor of 1 since it easier to maintain and will provide increased perf on a100 machines for the hot path (no dropout).

For both paths, updates to internal build system will be needed.

**Comment by jbschlosser** [comment]
> I think the main difference is that FAv3 doesn't support Dropout

Hm this seems like a big enough blocker to me :) It's basically saying that FA3 is not fully BC with FA2.

**Comment by tridao** [comment]
There's plan to add dropout to FA3, but that hasn't been a high priority since fewer people use dropout on attention nowadays. If BC is important we can add back dropout.

**Comment by WingsOfPanda** [comment]
I personally prefer to switch to fa3 as soon as possible (opt 1 ^^), but I’m currently working across multiple GPU platforms, including H200, A100, and L20 machines. From my testing, I’ve observed notable differences in BF16 precision between fa3 (optimized exclusively for H200) and fa2. Therefore, when deploying on A100, L20, or GPUs such as 5090 and 4090, I need to revert to fa2 to maintain precision consistency.

Given this, I think Option 2 is preferable because:
	1.	It offers greater flexibility for users who might prefer different features such as dropout or other configurations available in fa2.
	2.	Users can easily switch between fa2 and fa3, enabling rapid fine-tuning on H200 while preserving precision compatibility with lower-end GPUs like L20, A100, 5090, and 4090.

**Comment by tridao** [comment]
FA2 and FA3 should produce very similar numerical precision

**Comment by fno2010** [comment]
I vote for Option 2 at the current moment. If we can finally merge both FAv2 and FAv3 into a single coherent implementation, we can mark FAv2 as deprecated and move to Option 1 later. But considering FAv3 still has some incompatible issues right now (e.g., dropout), if we want a quick adoption, I think Option 2 is better.

**Comment by EduardDurech** [comment]
Any updates?

**Comment by WingsOfPanda** [comment]
Any updates? We work with very high-resolution, high-dimensional medical images using ViT models, so compute is always our biggest bottleneck. It would be fantastic to see integration for running FA3 to further accelerate our training pipelines~

**Comment by samsja** [comment]
I think having another backend for fa3 is fine for now, not sure if this is the blocker but not having fa3 in sdpa is a big blocker for many usecase (native torch CP need sdpa for example)

**Comment by drisspg** [comment]
Closing as we have since landed support: https://docs.pytorch.org/docs/2.12/generated/torch.nn.attention.activate_flash_attention_impl.html#torch.nn.attention.activate_flash_attention_impl

We have abi compat wheels here: https://download.pytorch.org/whl/flash-attn-3/

FA4 wheels can be found here: https://pypi.org/project/flash-attn-4/
